.github/actions/directories/action.yml

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 The Authors (see AUTHORS file)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: 'Guardian Directories'
+description: |-
+  Use this action to get a list of the Terraform directories to use with Guardian for your repository. Guardian will find and use all directories that contain a Terraform backend as an allowlist.
+
+inputs:
+  detect_changes:
+    description: 'Uses the list of changed files to determine the directories run Guardian commands for.'
+    required: false
+    default: 'false'
+  terraform_directory:
+    description: 'A subfolder to use as the root of all Terraform configuration for this repository.'
+    required: false
+    default: '.'
+  directories:
+    description: 'A CSV list of directories to run Guardian commands for.'
+    required: false
+    default: ''
+
+outputs:
+  dirs:
+    description: 'The directories to use as inputs to Guardian actions.'
+    value: ${{ steps.get_dirs.outputs.dirs }}
+
+runs:
+  using: composite
+  steps:
+    - name: 'Get Directories'
+      id: 'get_dirs'
+      shell: 'bash'
+      run: |-
+        TERRAFORM_GLOB=$(echo "./${{ inputs.terraform_directory }}/**/*.tf" | sed 's|././|./|g' | sed 's|//|/|g')
+        DIRS_WITH_BACKENDS=$(grep -E 'backend\s?".*"\s?{' $TERRAFORM_GLOB -l | xargs dirname | sed 's|^[\.?][/?]*||g' | sort | uniq)
```

Comment:
General request: can you put comments with your regexes to explain what they are doing? Mainly for future maintainers but it also aids in review.

Reply:
ack